src/rust/catnap/win/overlapped.rs

```diff
@@ -0,0 +1,774 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+//==============================================================================
+// Imports
+//==============================================================================
+
+use std::{
+    cell::{
+        Cell,
+        UnsafeCell,
+    },
+    marker::{
+        PhantomData,
+        PhantomPinned,
+    },
+    pin::Pin,
+};
+
+use windows::Win32::{
+    Foundation::{
+        CloseHandle,
+        ERROR_INVALID_HANDLE,
+        FALSE,
+        HANDLE,
+        INVALID_HANDLE_VALUE,
+        NTSTATUS,
+        WAIT_ABANDONED,
+        WAIT_TIMEOUT,
+        WIN32_ERROR,
+    },
+    Networking::WinSock::SOCKET,
+    System::IO::{
+        CreateIoCompletionPort,
+        GetQueuedCompletionStatusEx,
+        OVERLAPPED,
+        OVERLAPPED_ENTRY,
+    },
+};
+
+use crate::{
+    catnap::transport::error::translate_ntstatus,
+    runtime::{
+        fail::Fail,
+        scheduler::{
+            Yielder,
+            YielderHandle,
+        },
+    },
+};
+
+//======================================================================================================================
+// Structures
+//======================================================================================================================
+
+/// A structure to represent the results of an overlapped I/O operation.
+#[derive(Clone, Copy)]
+pub struct OverlappedResult {
+    /// The completion key for the operation, as set with `IoCompletionPort::associate`.
+    pub completion_key: usize,
+    /// Status result of the operation.
+    pub result: NTSTATUS,
+    /// Number of bytes transferred, if applicable.
+    pub bytes_transferred: u32,
+}
+
+/// Data required by the I/O completion port processor to process I/O completions.
+#[repr(C)]
+struct OverlappedCompletion {
+    /// OVERLAPPED must be first; the implementation casts between the outer structure and this type.
+    overlapped: OVERLAPPED,
+    /// If set, indicates a coroutine in waiting. Cleared by the I/O processor to signal completion. If unset when an
+    /// overlapped is dequeued from the completion port, the completion is abandoned.
+    yielder_handle: Option<YielderHandle>,
+    /// Set by the I/O processor to indicate overlapped result.
+    completion_key: usize,
+    /// A callback to free all resources associated with the completion for abandoned waits.
+    free: unsafe fn(*mut OVERLAPPED) -> (),
+    /// Ensure the data stays pinned since the OVERLAPPED must be pinned.
+    _marker: PhantomPinned,
+}
+
+/// The set of data which must live as long as the I/O operation, including any optional state from the caller.
+#[repr(C)]
+struct StatefulOverlappedCompletion<S> {
+    /// Inner data required by the completion processor; must be first to "downcast" to OVERLAPPED.
+    inner: OverlappedCompletion,
+    /// Caller state, encased in a cell for interior mutability.
+    state: UnsafeCell<S>,
+}
```

Comment:
If we also must ensure a specific layout on this, it may be worth to add some compile-time assertions.